include/sycl/algorithm/sort.hpp

```diff
@@ -173,6 +173,780 @@ void sequential_sort(cl::sycl::queue q, cl::sycl::buffer<T, 1, Alloc> buf,
+  typename bitonic_sort_base<T, U>::local_buffer_accessor_t m_localBuf;
+  const unsigned int m_stage;
+  const int mDir;
+};
+
+/**
+ * @brief
+ *
+ */
+template <typename T, unsigned U>
+class bitonic_sort_merge_last : public bitonic_sort_base<T, U> {
+public:
+  bitonic_sort_merge_last(
+      const typename bitonic_sort_base<T, U>::global_buffer_accessor_t
+          &globalBuf,
+      const typename bitonic_sort_base<T, U>::local_buffer_accessor_t &localBuf,
+      const int dir)
+      : m_globalBuf(globalBuf), m_localBuf(localBuf), mDir(dir) {}
+
+  /**
+   * @brief
+   *
+   */
+  void operator()(cl::sycl::nd_item<1> item) {
+    using namespace cl::sycl;
+#ifdef EMULATE_SHUFFLE_BUILTINS
+    using namespace emulated_shuffle_builtins;
+#endif
+
+    typedef typename bitonic_sort_base<T, U>::data_vec_type data_vec_type_;
+    typedef
+        typename bitonic_sort_base<T, U>::mask_op_vec_type mask_op_vec_type_;
+    typedef typename bitonic_sort_base<T, U>::relational_op_vec_type
+        relational_op_vec_type_;
+    const int vec_elems_per_thread_ =
+        bitonic_sort_base<T, U>::vec_elems_per_thread_::value;
+
+    unsigned int id, global_start, stride;
+    data_vec_type_ input1, input2, temp;
+    relational_op_vec_type_ comp;
+
+    mask_op_vec_type_ mask1(1, 0, 3, 2);
+    mask_op_vec_type_ mask2(2, 3, 0, 1);
+    mask_op_vec_type_ mask3(3, 2, 1, 0);
+
+    relational_op_vec_type_ add1(1, 1, 3, 3);
+    relational_op_vec_type_ add2(2, 3, 2, 3);
+    relational_op_vec_type_ add3(4, 5, 6, 7);
+
+    /* Determine location of data in global memory */
+    id = item.get_local_id(0);
+    global_start =
+        item.get_group(0) * item.get_local_range(0) * vec_elems_per_thread_ +
+        id;
+
+    /* Perform initial swap */
+    input1 = m_globalBuf[global_start];
+    input2 = m_globalBuf[global_start + item.get_local_range(0)];
+    comp = ((input1 < input2 ^ mDir) * 4 + add3);
+    m_localBuf[id] =
+        shuffle2(input1, input2, (comp).template as<mask_op_vec_type_>());
+    m_localBuf[id + item.get_local_range(0)] =
+        shuffle2(input2, input1, (comp).template as<mask_op_vec_type_>());
+
+    /* Perform bitonic merge */
+    for (stride = item.get_local_range(0) / vec_elems_per_thread_; stride > 1;
+         stride >>= 1) {
+      item.barrier(access::fence_space::local_space);
+      id = item.get_local_id(0) + (item.get_local_id(0) / stride) * stride;
+      VECTOR_SWAP(m_localBuf[id], m_localBuf[id + stride], mDir)
+    }
+    item.barrier(access::fence_space::local_space);
+
+    /* Perform final sort */
+    id = item.get_local_id(0) * vec_elems_per_thread_;
+    input1 = m_localBuf[id];
+    input2 = m_localBuf[id + 1];
+    temp = input1;
+    comp = ((input1 < input2 ^ mDir) * 4 + add3);
+    input1 = shuffle2(input1, input2, (comp).template as<mask_op_vec_type_>());
+    input2 = shuffle2(input2, temp, (comp).template as<mask_op_vec_type_>());
+    VECTOR_SORT(input1, mDir);
+    VECTOR_SORT(input2, mDir);
```

Comment:
By browsing some code that seems similar at different level accress different kernels, I wonder whether it could not be possible to have some even more generic code and code refactorizing...